File: src/Memories/Domain/Memory.php
<?php

namespace App\Memories\Domain;

use App\Categories\Domain\ForbiddenNameException;
use App\Memories\Domain\ForbiddenWords\ForbiddenWordChecker;
use App\Shared\Domain\Entity;
use App\Shared\Domain\Exceptions\CustomException;
use App\Users\Domain\UserId;
use DateTimeImmutable;

class Memory extends Entity
{
    private function __construct(
        private readonly MemoryId          $id,
        private readonly MemoryName        $name,
        private readonly MemoryType        $type,
        private readonly DateTimeImmutable $createdAt,
        private readonly MemoryCategories  $categories,
        private readonly UserId            $userId,
        private ?MemoryContent             $content,
        private ?DateTimeImmutable         $modifiedAt,
    )
    {
    }

    /**
     * @throws ForbiddenNameException
     */
    public static function create(
        MemoryId             $id,
        MemoryName           $name,
        MemoryType           $type,
        DateTimeImmutable    $createdAt,
        MemoryCategories     $categories,
        UserId               $userId,
        MemoryContent|null   $content,
        ForbiddenWordChecker $forbiddenWordChecker,
    ): self
    {
        $forbiddenWordChecker->assert($content?->value);

        $memory = new self($id, $name, $type, $createdAt, $categories, $userId, $content, null);
        $memory->dispatchCreated();

        return $memory;
    }

    /**
     * @throws CustomException
     */
    public static function fromPrimitives(
        string            $id,
        string            $name,
        int               $type,
        DateTimeImmutable $createdAt,
        array             $categories,
        string            $userId,
        ?string           $content
    ): self
    {
        return new self(
            MemoryId::fromValue($id),
            MemoryName::fromValue($name),
            MemoryType::fromValue($type),
            $createdAt,
            MemoryCategories::fromArray($categories),
            UserId::fromValue($userId),
            $content ? MemoryContent::fromValue($content) : null,
            null,
        );
    }

    private function dispatchCreated(): void
    {
        $this->events[] = new MemoryCreated($this->id->value);
    }

    /**
     * @throws CustomException
     */
    public function updateContent(MemoryContent $content, ForbiddenWordChecker $checker): void
    {
        $checker->assert($content->value);

        $this->content = $content;
        $this->modifiedAt = new DateTimeImmutable();

        $this->events[] = new ContentUpdated($this->id->value);
    }

    public function getId(): MemoryId
    {
        return $this->id;
    }

    public function getName(): MemoryName
    {
        return $this->name;
    }

    public function getType(): MemoryType
    {
        return $this->type;
    }

    public function getCreatedAt(): DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function getCategories(): MemoryCategories
    {
        return $this->categories;
    }

    public function getUserId(): UserId
    {
        return $this->userId;
    }

    public function getContent(): ?MemoryContent
    {
        return $this->content;
    }

    public function getModifiedAt(): ?DateTimeImmutable
    {
        return $this->modifiedAt;
    }
}

File: src/main/kotlin/com/zaleort/second_brain/Memories/Domain/Memory/Memory.kt
package com.zaleort.second_brain.Memories.Domain.Memory

import com.zaleort.second_brain.Users.Domain.User.UserId
import java.time.Instant

class Memory private constructor(
    val id: MemoryId,
    val userId: UserId,
    var title: MemoryTitle,
    var content: MemoryContent,
    var type: MemoryType,
    var tags: List<MemoryTag>,
    val createdAt: Instant,
    var updatedAt: Instant?,
) {
    fun update(
        title: MemoryTitle,
        content: MemoryContent,
        type: MemoryType,
        tags: List<MemoryTag>,
    ) {
        this.title = title
        this.content = content
        this.type = type
        this.tags = tags
        this.updatedAt = Instant.now()
    }

    companion object {
        fun create(
            id: MemoryId,
            userId: UserId,
            title: MemoryTitle,
            content: MemoryContent,
            type: MemoryType,
            tags: List<MemoryTag>,
        ): Memory {
            return Memory(
                id = id,
                userId = userId,
                title = title,
                content = content,
                type = type,
                tags = tags,
                createdAt = Instant.now(),
                updatedAt = null
            )
        }

        fun fromPrimitives(
            id: String,
            userId: String,
            title: String,
            content: String,
            type: Int,
            tags: List<Map<String, String?>>,
            createdAt: String,
            updatedAt: String?,
        ): Memory {
            return Memory(
                id = MemoryId(id),
                userId = UserId(userId),
                title = MemoryTitle(title),
                content = MemoryContent(content),
                type = MemoryType(type),
                tags = tags.map {
                    MemoryTag(
                        it["id"] ?: "",
                        it["name"] ?: "",
                        it["color"] ?: "",
                    )
                },
                createdAt = Instant.parse(createdAt),
                updatedAt = updatedAt?.let { Instant.parse(it) }
            )
        }
    }
}